src/gdrapi.c

```diff
@@ -537,59 +537,70 @@ static inline int ptr_is_aligned(const void *ptr, unsigned powof2)
 
 static int gdr_copy_to_mapping_internal(void *map_d_ptr, const void *h_ptr, size_t size, int wc_mapping)
 {
-    if (first_time) {
-        gdr_init_cpu_flags();
-    }
-
     do {
+        // For very small sizes and aligned pointers, we use simple store.
+        if (unlikely(size == 0)) {
+            goto out;
+        } else if (size == 1) {
+            WRITE_ONCE(*(uint8_t *)map_d_ptr, *(uint8_t *)h_ptr);
+            goto do_fence;
+        } else if (size == 2 && ptr_is_aligned(map_d_ptr, 2) && ptr_is_aligned(h_ptr, 2)) {
+            WRITE_ONCE(*(uint16_t *)map_d_ptr, *(uint16_t *)h_ptr);
+            goto do_fence;
+        } else if (size == 4 && ptr_is_aligned(map_d_ptr, 4) && ptr_is_aligned(h_ptr, 4)) {
+            WRITE_ONCE(*(uint32_t *)map_d_ptr, *(uint32_t *)h_ptr);
+            goto do_fence;
+        } else if (size == 8 && ptr_is_aligned(map_d_ptr, 8) && ptr_is_aligned(h_ptr, 8)) {
+            WRITE_ONCE(*(uint64_t *)map_d_ptr, *(uint64_t *)h_ptr);
+            goto do_fence;
+        }
+
         // pick the most performing implementation compatible with the platform we are running on
         // NOTE: write fences are included in functions below
         if (has_avx) {
             assert(wc_mapping);
             gdr_dbgc(1, "using AVX implementation of gdr_copy_to_bar\n");
             memcpy_uncached_store_avx(map_d_ptr, h_ptr, size);
-            break;
+            goto out;
         }
         if (has_sse) {
             assert(wc_mapping);
             gdr_dbgc(1, "using SSE implementation of gdr_copy_to_bar\n");
             memcpy_uncached_store_sse(map_d_ptr, h_ptr, size);
-            break;
+            goto out;
         }
 
         // on POWER, compiler/libc memcpy is not optimal for MMIO
-        // 64bit stores are not better than 32bit ones, so we prefer the latter
+        // 64bit stores are not better than 32bit ones, so we prefer the latter.
         // NOTE: if preferred but not aligned, a better implementation would still try to
-        // use byte sized stores to align map_d_ptr and h_ptr to next word
+        // use byte sized stores to align map_d_ptr and h_ptr to next word.
+        // NOTE2: unroll*_memcpy and memcpy do not include fencing. So, we may need the fencing below.
```

Comment:
I believe that after moving the fence out of the do-while block, the "SO, we may..." part of the comment does not make much sense anymore.

Reply:
Removed.